File: Graphs/994. Rotting Oranges/solution.py
from queue import Queue
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        xdir= [-1,0,1,0]
        ydir = [0, 1, 0,-1]
        n, m = len(grid), len(grid[0])
        q = Queue()
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 2:
                    q.put((i,j))
        count = 0
        while not q.empty():
            s = q.qsize()
            flag = 0
            for i in range(s):
                x = q.get()
                for d in range(4):
                    ni = x[0] + xdir[d]
                    nj = x[1] + ydir[d]
                    if ni >= 0 and ni < n and nj >=0 and nj < m and grid[ni][nj] == 1:
                        flag = 1
                        grid[ni][nj] = 2
                        q.put((ni, nj))
            if flag == 1:
                count += 1
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    return -1
        return count

File: Graphs/994. Rotting Oranges/solution.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& v) {
        int x4[] = {-1,0,1,0};
        int y4[] = {0, 1, 0,-1};
        int days = 0;
        int n = v.size();
        int m = v[0].size();
        queue<pair<int, int>> q;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(v[i][j] == 2){
                    q.push({i,j});
                }
                
            }
        }
        while(!q.empty()){
            int z = q.size();
            int flag = 0;
            for(int i = 0; i < z; i++){
                pair<int, int> p = q.front();
                q.pop();
                for(int d= 0; d < 4; d++){
                    int ni = p.first + x4[d];
                    int nj = p.second + y4[d];
                    if(ni >= 0 && ni < n && nj >= 0 && nj < m && v[ni][nj] == 1){
                        flag = 1;
                        q.push({ni, nj});
                        v[ni][nj] = 2;
                    }
                }

            }
            if(flag == 1)
                days += 1;
        }

        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(v[i][j] == 1){
                    days = -1;
                }
            }
        }

        return days;
        
    }
};
